filament/backend/src/opengl/platforms/CocoaExternalImage.mm

```diff
@@ -0,0 +1,236 @@
+    glSamplerParameteri(sampler, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glSamplerParameteri(sampler, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
+
+    GLint status;
+
+    vertexShader = glCreateShader(GL_VERTEX_SHADER);
+    glShaderSource(vertexShader, 1, &s_vertex, nullptr);
+    glCompileShader(vertexShader);
+    glGetShaderiv(vertexShader, GL_COMPILE_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    glShaderSource(fragmentShader, 1, &s_fragment, nullptr);
+    glCompileShader(fragmentShader);
+    glGetShaderiv(fragmentShader, GL_COMPILE_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    program = glCreateProgram();
+    glAttachShader(program, vertexShader);
+    glAttachShader(program, fragmentShader);
+    glLinkProgram(program);
+    glGetProgramiv(program, GL_LINK_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    // Save current program state.
+    GLint currentProgram;
+    glGetIntegerv(GL_CURRENT_PROGRAM, &currentProgram);
+
+    glUseProgram(program);
+    GLint samplerLoc = glGetUniformLocation(program, "rectangle");
+    glUniform1i(samplerLoc, 0);
+
+    // Restore state.
+    glUseProgram(currentProgram);
+}
+
+CocoaExternalImage::SharedGl::~SharedGl() noexcept {
+    glDeleteSamplers(1, &sampler);
+    glDetachShader(program, vertexShader);
+    glDetachShader(program, fragmentShader);
+    glDeleteShader(vertexShader);
+    glDeleteShader(fragmentShader);
+    glDeleteProgram(program);
+}
+
+CocoaExternalImage::CocoaExternalImage(const CVOpenGLTextureCacheRef textureCache,
+        const SharedGl &sharedGl) noexcept : mSharedGl(sharedGl), mTextureCache(textureCache) {
+    glGenFramebuffers(1, &mFBO);
+    CHECK_GL_ERROR(utils::slog.e)
+}
+
+CocoaExternalImage::~CocoaExternalImage() noexcept {
+    glDeleteFramebuffers(1, &mFBO);
+    release();
+}
+
+bool CocoaExternalImage::set(CVPixelBufferRef image) noexcept {
+    // Release references to a previous external image, if we're holding any.
+    release();
+
+    if (!image) {
+        return false;
+    }
+
+    OSType formatType = CVPixelBufferGetPixelFormatType(image);
+    ASSERT_POSTCONDITION(formatType == kCVPixelFormatType_32BGRA,
+            "macOS external images must be 32BGRA format.");
+
+    // The pixel buffer must be locked whenever we do rendering with it. We'll unlock it before
+    // releasing.
+    UTILS_UNUSED_IN_RELEASE CVReturn lockStatus = CVPixelBufferLockBaseAddress(image, 0);
+    assert_invariant(lockStatus == kCVReturnSuccess);
+
+    mImage = image;
+    mTexture = createTextureFromImage(image);
+    mRgbaTexture = encodeCopyRectangleToTexture2D(CVOpenGLTextureGetName(mTexture),
+            CVPixelBufferGetWidth(image), CVPixelBufferGetHeight(image));
+    CHECK_GL_ERROR(utils::slog.e)
+
+    return true;
+}
+
+GLuint CocoaExternalImage::getGlTexture() const noexcept {
+    if (mEncodedToRgba) {
```

Comment:
Do we even need the `mEncodedToRgba` flag? Seems that after `set` is called, `mRgbaTexture` will be valid. Before set is called, neither `mRgbaTexture` nor `mTexture` are valid, so maybe we should return 0 or throw an exception in that case.

Reply:
Yes, that makes sense. Your feedback will be taken into account.